Reusable actions

iaindillingham opened the pull request:
Behold, reusable actions! 🎉 

* This is part one; for part two, see #228.
* For a discussion of how actions are hosted in GitHub, see #229. (TL;DR we will host actions in a new GitHub organisation called `opensafely-actions`.)
* For more information about the nature of reusable actions, see https://github.com/opensafely-core/action-packager/issues/1#issuecomment-874200067.

What's in this PR?

* Two small commits that move code for parsing YAML and validating the project file (*project.yaml*)
* One small commit that parses and validates the action file (*action.yaml*). @iaindillingham needs to revisit this commit, as he needs to think about what to `raise`; if an action developer has created an invalid action file, then there's only so much a study developer can do 🤷🏻‍♂️ 
* One small commit that gets the "binary" for a reusable action, if a reusable action has a binary. Python reusable actions do (`python`); R and Stata reusable actions don't
* One big commit that handles reusable actions - rewriting the action specification before it's saved (in the sync loop), to take advantage of existing functionality that fetches a Git repository and copies it to a volume. There is a small change when running a job (in the run loop), which introduces a switch to ensure that reusable actions are fetched when running in local run mode.

Closes #227

evansd reviewed (changes requested):
This looks good, and definitely like the right approach. There are a couple of issues to consider first though.

One issue is that this won't work using `opensafely run` as it stands because local run mode sets `config.GIT_REPO_DIR = None` deliberately as previously there shouldn't be any git operations in local run mode. Making this work is just a question of setting it so _something_, the question is what. One option is to use a temporary directory like we do for `LOG_DIR` which gets thrown away at the end of the run. The other option is to use a directory in something like $XDG_CACHE_HOME (or its Windows equivalent) so it persists between runs. @bloodearnest had some concerns here about force-pushes causing us problems, but I don't think that will be an issue because we use these directories as essentially big bags of commits, and do all the ref resolution manually ourselves. (We even add our own [atomicity](https://github.com/opensafely-core/job-runner/pull/209) layer!)

The second issue is that `get_action_specification` gets called inside a transaction block in `create_or_update_jobs.py` and now it ends up talking to git, which is potentially quite a long running operation (especially in production where it sometimes randomly fails and we have to retry). But because of the transaction SQLite will lock the entire database for both reads and writes during this period, which will mean that the run loop stalls. Ideally we want to be doing all our network IO before we enter the transaction. It might be possible to address this by handling all the git stuff and `action.yaml` parsing in an initial pass over project.yaml  to add the relevant repo_urls and SHAs before entering the transaction. The downside to that would be that we end up fetching code for actions that we aren't actually going to run this time. But that feels like another point in favour of using a persistent `GIT_DIR` as this would be quite a cheap operation after the first time it had been run.

iaindillingham commented on jobrunner/manage_jobs.py:
```diff
@@ -125,7 +125,7 @@ def create_and_populate_volume(job):
     # these directories get created when we copy in the code
     extra_dirs = set(Path(filename).parent for filename in input_files.keys())
-    if config.LOCAL_RUN_MODE:
+    if config.LOCAL_RUN_MODE and not job.is_action_reusable:
```
Is `job.commit` `"none"` and not `None` because of how values are written to SQLite? Or because `JobRequest.commit` should be `str` rather than, say, `Optional[str]`? Or something else?

iaindillingham replied:
I'm wondering whether we can do without magic strings.

evansd replied:
Ah, now I come to look at it I think the only reason I ended up using `"none"` was because of this line here:
https://github.com/opensafely-core/job-runner/blob/64b037a96f4bf19914e6a78d4c6815bac2c1945e/jobrunner/create_or_update_jobs.py#L94

But I think we can just change that to:
```py
if not config.LOCAL_RUN_MODE and not job_request.commit
```

I think both `None` and `""` would work here as the database column is NULLable.

Agree that it would be better to do without magic strings if we can.

iaindillingham replied:
I'm going to push two commits. One replaces `"none"` with `None`. One removes `Job.is_action_reusable` and instead tests the value of `Job.commit`. I think both resolve this issue.

evansd replied:
:+1: I'd be inclined to just test for falseyness though, rather than explicitly checking `is None`. I'm not 100% sure the None will round-trip successfully through the db and I don't think it gains us anything to not handle the empty string as well. If you're confident it works though then happy to leave as is.

iaindillingham replied:
I'm going to push a commit that tests for falseyness, as I'm also not sure what the DB does.

iaindillingham commented on jobrunner/project.py:
```diff
@@ -234,6 +237,35 @@ def get_action_specification(project, action_id):
+        # Modify run_args
+        # ["action:tag", "arg", ...]
+
+        run_args[0] = action["runtime"]
+        # ["runtime:tag", "arg", ...]
+
+        run_args.insert(1, action["entrypoint"])
+        # ["runtime:tag", "entrypoint", "arg", ...]
+
+        binary = get_binary_for_reusable_action(action)
+        if binary:
+            run_args.insert(1, binary)
+            # ["runtime:tag", "binary", "entrypoint", "arg", ...]
```
They need not be; we could instead ask that *action.yaml* contains one property, possibly called `run`. For example:

```yaml
run: python:latest python reusable_action/main.py
```

We'd replace `run_args[0]` with `action["run"]` on line 253 above. Is that what you're suggesting? If so, then I'll pop a message in Slack, as I can't remember why we separated the elements. This approach would be better, though: it would make writing a reusable action even more like writing an action.

evansd replied:
Yes, effectively. Although I think to get the argument separation working it would need to be:
```py
run_args[:1] = shlex.split(action["run"])
```

iaindillingham replied:
I'm going to push a commit that replaces `runtime`/`entrypoint` with `run`. We no longer need to assemble a replacement for `run_args[0]` and can remove `get_binary_for_reusable_action`. I don't think we need to separate the arguments, though. Possibly the word "action" is confusing, here?

```python
action = {"run": "reusable-action:latest --output-format=png"}  # From project.yaml
run_args = shlex.split(action["run"])
run_args  # ['reusable-action:latest', '--output-format=png']

action_config = {"run": "python:latest python reusable_action/main.py"}  # From action.yaml

run_args[0] = action_config["run"]  # No argument separation!
run_args  # ['python:latest python reusable_action/main.py', '--output-format=png']

action["run"] = " ".join(run_args)
action  # {'run': 'python:latest python reusable_action/main.py --output-format=png'}
```

iaindillingham commented:
WRT to calling `opensafely run`, I'm going to push a commit that introduces a `temp_git_repo_dir` argument to `jobrunner.local_run.create_and_run_jobs`. This is set to `project_dir / ".reusable_actions"` in `main` and won't persist between runs (so it's similar to `temp_log_dir`).

I'm not sure how I'd test this, though. I guess by adding a reusable action to the full project fixture? (*tests/fixtures/full_project/project.yaml*).

iaindillingham commented:
I should probably have addressed ☝🏻 after addressing your comment about `get_action_specification`, but I can revisit now that I understand ☝🏻 better. (And also, I don't think that fetching a reusable action's git repo twice will be noticeable, given that jobs can be long-running.)

iaindillingham commented:
WRT to the transaction block, I'm going to push two commits that bracket moving where we handle reusable actions from `get_action_specification` (called inside the transaction block) to `parse_and_validate_project_file` (called outside the transaction block). These...

* remove code introduced in a previous commit. When we're done with this PR, I'll rebase to give a clear history.
* don't address your comments. More commits to follow!

Don't feel you have to 👀 this PR as I'm working through your comments. I don't want to resolve them/close the PR until I've addressed them, but appreciate that GitHub can be chatty.

iaindillingham commented on jobrunner/project.py:
```diff
@@ -234,6 +237,35 @@ def get_action_specification(project, action_id):
+        # Fetch the reusable action
+        repo_url = f"{config.ACTIONS_GITHUB_ORG}/{image}"
+        commit_sha = git.get_sha_from_remote_ref(repo_url, tag)
+        action_file = git.read_file_from_repo(repo_url, commit_sha, "action.yaml")
```
I'm going to push a commit that catches the `GitError`s and consolidates this with the contents of `parse_and_validate_action_file`.

`<` 🦆  `>`
I'm in two minds about logging, though: should we add more `log.exception` calls to `git.get_sha_from_remote_ref` and `git.read_file_from_repo`? If so, then should we do this here or in another PR?

I guess it depends on who is responsible for telling a reusable action developer that their reusable action isn't working and why. We will have some, but not all, of the "why" information in our logs. The study developer won't have any of this information. All they will know is that the reusable action isn't working. Perhaps logging is something we should revisit in another PR after all, when a few reusable actions are in production.
`</` 🦆 `>`

iaindillingham commented on jobrunner/project.py:
```diff
@@ -234,6 +237,35 @@ def get_action_specification(project, action_id):
+    image, tag = run_args[0].split(":")
+    if image not in config.ALLOWED_IMAGES:
+        # Fetch the reusable action
+        repo_url = f"{config.ACTIONS_GITHUB_ORG}/{image}"
```
I'm going to push a commit that moves the scheme and netloc to the config and calls `create_or_update_jobs.validate_repo_url` on `repo_url`.

iaindillingham commented on jobrunner/project.py:
```diff
@@ -234,6 +237,35 @@ def get_action_specification(project, action_id):
+    if image not in config.ALLOWED_IMAGES:
+        # Fetch the reusable action
+        repo_url = f"{config.ACTIONS_GITHUB_ORG}/{image}"
+        commit_sha = git.get_sha_from_remote_ref(repo_url, tag)
```
I'm going to push a commit that calls `validate_branch_and_commit` to check that we can reach the commit from the repo's `main` branch. I'll update the [current proposal](https://github.com/opensafely-core/action-packager/issues/1#issuecomment-874200067) to document that a reusable action's release branch must be the repo's `main` branch.

evansd commented on jobrunner/config.py:
```diff
@@ -110,9 +110,10 @@ class ConfigException(Exception):
 ALLOWED_GITHUB_ORGS = (
     os.environ.get("ALLOWED_GITHUB_ORGS", "opensafely").strip().split(",")
-)
+) + [ACTIONS_GITHUB_ORG]
```
I can see exactly why you'd want to do this, but unfortunately I don't think we can because we don't want to allow studies to be run from the `opensafely-actions` org. Action code has a slightly different risk profile from study code as it can only work on data that's already been extracted; it can't invoke cohortextractor to pull out more data. So we want to keep studies restricted to just the `opensafely` org and have a separate org for actions.

iaindillingham replied:
I've squashed the commits on this branch, but I'm going to address this by not touching `ALLOWED_GITHUB_ORGS`.

evansd commented on jobrunner/manage_jobs.py:
```diff
@@ -125,7 +125,7 @@ def create_and_populate_volume(job):
     # these directories get created when we copy in the code
     extra_dirs = set(Path(filename).parent for filename in input_files.keys())
-    if config.LOCAL_RUN_MODE and job.commit is None:
+    if config.LOCAL_RUN_MODE and not job.commit:
```
Do we want to make the equivalent change to the conditional in `create_or_update_jobs.py`?

iaindillingham replied:
Yup, this is consistent after squashing commits and tidying up.

evansd commented on jobrunner/create_or_update_jobs.py:
```diff
@@ -188,8 +188,8 @@ def recursively_add_jobs(job_request, project, action, force_run_actions):
-        repo_url=action_spec.repo_url if action_spec.repo_url else job_request.repo_url,
-        commit=action_spec.commit_sha if action_spec.commit_sha else job_request.commit,
+        repo_url=job_request.repo_url,
+        commit=job_request.commit,
```
I don't quite get how this works now. Don't we still need to pass the repo_url and commit through from the action?

It might be worth adding a reusable action step to the mammoth integration test just to make sure all the parts work together (which I think would have caught this). I agree it's a slightly intimidating giant of a test though, so happy to help out there.

iaindillingham replied:
After squashing commits and tidying up, there is very little change to *jobrunner/create_or_update_jobs.py*. I think it would be a good idea to close this PR, force pushing, and opening a new PR. After a -- hopefully quick -- review, then I will add to the integration test. How does that sound to you?

evansd replied:
Sounds good. Just include a link back to this PR in the new PR description for future github archaeologists

evansd reviewed (commented):
Thanks Iain. Looking good. I had a few comments and questions below but hopefully nothing requiring too much extra work!
